runtime/src/attestations.rs

```diff
@@ -0,0 +1,159 @@
+	traits::Get,
+};
+
+use inherents::{ProvideInherent, InherentData, RuntimeString, MakeFatalError, InherentIdentifier};
+use system::ensure_none;
+
+/// Parachain blocks included in a recent relay-chain block.
+#[derive(Encode, Decode)]
+pub struct IncludedBlocks<T: Trait> {
+	/// The actual relay chain block number where blocks were included.
+	pub actual_number: T::BlockNumber,
+	/// The session index at this block.
+	pub session: SessionIndex,
+	/// The randomness seed at this block.
+	pub random_seed: [u8; 32],
+	/// Hashes of the parachain candidates included at this block.
+	pub para_blocks: Vec<Hash>,
+}
+
+/// Attestations kept over time on a parachain block.
+#[derive(Encode, Decode)]
+pub struct BlockAttestations<T: Trait> {
+	receipt: CandidateReceipt,
+	valid: Vec<T::AccountId>, // stash account ID of voter.
+	invalid: Vec<T::AccountId>, // stash account ID of voter.
+}
+
+/// Additional attestations on a parachain block, after it was included.
+#[derive(Encode, Decode, Clone, PartialEq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct MoreAttestations;
+
+pub trait Trait: session::Trait {
+	/// How many blocks ago we're willing to accept attestations for.
+	type AttestationPeriod: Get<Self::BlockNumber>;
+
+	/// Get a list of the validators' underlying identities.
+	type ValidatorIdentities: Get<Vec<Self::AccountId>>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Attestations {
+		/// A mapping from modular block number (n % AttestationPeriod)
+		/// to session index and the list of candidate
+		/// hashes.
+		pub RecentParaBlocks: map T::BlockNumber => Option<IncludedBlocks<T>>;
+
+		/// Attestations on a recent parachain block.
+		pub ParaBlockAttestations: map (T::BlockNumber, Hash) => Option<BlockAttestations<T>>;
+
+		// Did we already have more attestations included in this block?
+		DidUpdate: bool;
+	}
+}
+
+decl_module! {
+	/// Parachains module.
+	pub struct Module<T: Trait> for enum Call where origin: <T as system::Trait>::Origin {
+		/// Provide candidate receipts for parachains, in ascending order by id.
+		fn more_attestations(origin, _more: MoreAttestations) -> Result {
+			ensure_none(origin)?;
+			ensure!(!<DidUpdate>::exists(), "More attestations can be added only once in a block.");
+
+			Ok(())
+		}
+
+		fn on_finalize(_n: T::BlockNumber) {
+			<Self as Store>::DidUpdate::kill();
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Update recent candidates to contain the already-checked parachain candidates.
+	pub(crate) fn note_included(heads: &[AttestedCandidate], para_blocks: IncludedBlocks<T>) {
+		let attestation_period = T::AttestationPeriod::get();
+		let mod_num = para_blocks.actual_number % attestation_period;
+
+		// clear old entry that was in this place.
+		if let Some(old_entry) = <RecentParaBlocks<T>>::take(&mod_num) {
+			for old_para_block in old_entry.para_blocks {
+				<ParaBlockAttestations<T>>::remove(&(old_entry.actual_number, old_para_block));
```

Comment:
An initial glance at this logic makes me feel like you could use a double_map here which provides an API to remove all values under the first key (`old_entry.actual_number`) rather than having to do this loop.

Reply:
nice!